### PR TITLE
add nullness handling to storage and its implementations

### PR DESCRIPTION
**Pull request by maggu2810**
A value of a key could be potential null. It depends on the respective implementation if it differs between a existing key that value is null or a nullable value states that the key does not exist.

Do you agree?

### CODE REVIEW

File: bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/internal/JsonStorage.java [hunk 1]
```diff
@@ -21,6 +21,8 @@
 import java.util.concurrent.ConcurrentHashMap;
 
 import org.eclipse.jdt.annotation.NonNull;
```
Comment (triller-telekom) [inline bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/internal/JsonStorage.java]: There is still @NonNull in use in this class, which should be superseded by @NonNullByDefault.

File: bundles/storage/org.eclipse.smarthome.storage.mapdb/src/main/java/org/eclipse/smarthome/storage/mapdb/internal/MapDbStorage.java [hunk 2]
```diff
@@ -9,10 +9,13 @@
 
 import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
 import java.util.HashSet;
 import java.util.Map;
 
 import org.eclipse.jdt.annotation.NonNull;
```
Comment (triller-telekom) [inline bundles/storage/org.eclipse.smarthome.storage.mapdb/src/main/java/org/eclipse/smarthome/storage/mapdb/internal/MapDbStorage.java]: There is still @NonNull in use in this class, which should be superseded by @NonNullByDefault.

File: bundles/test/org.eclipse.smarthome.test/src/main/groovy/org/eclipse/smarthome/test/storage/VolatileStorage.java [hunk 3]
```diff
@@ -12,6 +12,8 @@
 import java.util.concurrent.ConcurrentHashMap;
 
 import org.eclipse.jdt.annotation.NonNull;
```
Comment (triller-telekom) [inline bundles/test/org.eclipse.smarthome.test/src/main/groovy/org/eclipse/smarthome/test/storage/VolatileStorage.java]: There is still @NonNull in use in this class, which should be superseded by @NonNullByDefault.

**Comment by maggu2810** [comment]
> I agree that a `Map` should by default be annotated as `Map<@NonNull K, @Nullable V)`. But I vaguely remember that some maps do accept `null` as a key?

Sorry, I don't understand. You agree to which comment of mine?
And do you point to a Map usage of the Map interface?

**Comment by triller-telekom** [comment]
I was referring to:

> A value of a key could be potential null. It depends on the respective implementation if it differs between a existing key that value is null or a nullable value states that the key does not exist.
Do you agree?

And I was pointing to wherever implementations of the `Map` interface are used.

And for our `Storage` interface I think it also makes sense to only allow @NonNull keys but @Nullable values.

**Comment by maggu2810** [comment]
> And for our `Storage` interface I think it also makes sense to only allow @NonNull keys but @Nullable values.

Agree.

But this is all about our `Storage` interface that is completely independent of a `Map` interface.
If an implementation uses a `Map` it is up to the implementation if it choose to use one that allows nullable keys or one that does not support it.
So, I was confused that you "agree that a `Map`" but I didn't refer to a map in my comment.

**Pull request merged**